Lib/urllib/request.py

```diff
@@ -2492,33 +2492,34 @@ def real_close(self):
 # Proxy handling
 def getproxies_environment():
     """Return a dictionary of scheme -> proxy server URL mappings.
-
     Scan the environment for variables named <scheme>_proxy;
     this seems to be the standard convention.  If you need a
     different way, you can pass a proxies dictionary to the
     [Fancy]URLopener constructor.
-
     """
-    proxies = {}
     # in order to prefer lowercase variables, process environment in
     # two passes: first matches any, second pass matches lowercase only
-    for name, value in os.environ.items():
-        name = name.lower()
-        if value and name[-6:] == '_proxy':
-            proxies[name[:-6]] = value
+
+    # select only environment variables which end in (after making lowercase) _proxy
+    candidate_names = [name for name in os.environ.keys() if len(name)>5 and name[-6]=='_'] # fast selection of candidates
+    environment = [(name, os.environ[name], name.lower()) for name in candidate_names if name[-6:].lower()=='_proxy']
+
+    proxies = {}
+    for name, value, name_lower in environment:
+        if value and name_lower[-6:] == '_proxy':
+            proxies[name_lower[:-6]] = value
```

Comment:
There probably won't be lots of env vars with underscore sixth-from-last, so it likely won't make a big difference, but I think there's really no need for three separate loops here. We can get the same benefits with a single loop and just better ordering of checks/operations. Also, every use of the lowered name wants to strip off final "_proxy", so we can do that early.
```suggestion
    proxies = {}
    environment = []
    for name in os.environ.keys():
        # fast screen underscore position before more expensive case-folding
        if len(name) > 5 and name[-6] == "_" and name[-5:].lower() == "proxy":
            value = os.environ[name]
            proxy_name = name[:-6].lower()
            environment.append((name, value, proxy_name))
            if value:
                proxies[proxy_name] = value
```